Game/Source/GDKTestGyms/BenchmarkGymGameMode.cpp

```diff
@@ -0,0 +1,292 @@
+		{
+			PlayerDensity = FCString::Atoi(*PlayerDensityString);
+		}
+
+		if (USpatialWorkerFlags::GetWorkerFlag(TEXT("total_npcs"), TotalNPCsString))
+		{
+			TotalNPCs = FCString::Atoi(*TotalNPCsString);
+		}
+	}
+
+	NumPlayerClusters = FMath::CeilToInt(TotalPlayers / (float)PlayerDensity);
+}
+
+void ABenchmarkGymGameMode::ClearExistingSpawnPoints()
+{
+	UGameplayStatics::GetAllActorsOfClass(GetWorld(), APlayerStart::StaticClass(), SpawnPoints);
+	for (AActor* SpawnPoint : SpawnPoints)
+	{
+		SpawnPoint->Destroy();
+	}
+}
+
+void ABenchmarkGymGameMode::GenerateGridSettings(int DistBetweenPoints, int NumPoints, int& NumRows, int& NumCols, int& MinRelativeX, int& MinRelativeY)
+{
+	if (NumPoints <= 0)
+	{
+		UE_LOG(LogBenchmarkGym, Warning, TEXT("Generating grid settings with non-postive number of points (%d)"), NumPoints);
+		NumRows = 0;
+		NumCols = 0;
+		MinRelativeX = 0;
+		MinRelativeY = 0;
+		return;
+	}
+
+	NumRows = FMath::RoundToInt(FMath::Sqrt(NumPoints));
+	NumCols = FMath::CeilToInt(NumPoints / (float)NumRows);
+	const int GridWidth = (NumCols - 1) * DistBetweenPoints;
+	const int GridHeight = (NumRows - 1) * DistBetweenPoints;
+	MinRelativeX = FMath::RoundToInt(-GridWidth / 2.0);
+	MinRelativeY = FMath::RoundToInt(-GridHeight / 2.0);
+}
+
+void ABenchmarkGymGameMode::GenerateSpawnPointClusters(int NumClusters)
+{
+	const int DistBetweenClusterCenters = 40000; // 400 meters, in Unreal units.
+	int NumRows, NumCols, MinRelativeX, MinRelativeY;
+	GenerateGridSettings(DistBetweenClusterCenters, NumClusters, NumRows, NumCols, MinRelativeX, MinRelativeY);
+
+	UE_LOG(LogBenchmarkGym, Log, TEXT("Creating player cluster grid of %d rows by %d columns"), NumRows, NumCols);
+	for (int i = 0; i < NumClusters; i++)
+	{
+		const int Row = i % NumRows;
+		const int Col = i / NumRows;
+
+		const int ClusterCenterX = MinRelativeX + Col * DistBetweenClusterCenters;
+		const int ClusterCenterY = MinRelativeY + Row * DistBetweenClusterCenters;
+
+		GenerateSpawnPoints(ClusterCenterX, ClusterCenterY, PlayerDensity);
+	}
+}
+
+void ABenchmarkGymGameMode::GenerateSpawnPoints(int CenterX, int CenterY, int SpawnPointsNum)
+{
+	// Spawn in the air above terrain obstacles (Unreal units).
+	const int Z = 300;
+
+	const int DistBetweenSpawnPoints = 300; // In Unreal units.
+	int NumRows, NumCols, MinRelativeX, MinRelativeY;
+	GenerateGridSettings(DistBetweenSpawnPoints, SpawnPointsNum, NumRows, NumCols, MinRelativeX, MinRelativeY);
+
+	UWorld* World = GetWorld();
+	if (World == nullptr)
+	{
+		UE_LOG(LogBenchmarkGym, Error, TEXT("Cannot spawn spawnpoints, world is null"));
+		return;
+	}
+
+	const int WalkRadius = 200; // In Unreal units. More than actual radius, just to be sure.
+	const int MaxXDistance = (NumCols - 1) * DistBetweenSpawnPoints + 2 * WalkRadius;
+	const int MaxYDistance = (NumRows - 1) * DistBetweenSpawnPoints + 2 * WalkRadius;
+	const int MaxDistanceBetweenPlayers = FMath::Sqrt(FMath::Square(MaxXDistance) + FMath::Square(MaxYDistance));
```

Comment:
Nit: don't have to do this at all.
You could do: 

const int MaxSqrDistanceBetweenPlayers = FMath::Square(MaxXDistance) + FMath::Square(MaxYDistance);
if (MaxSqrDistanceBetweenPlayers >= PlayerCheckoutRadius*PlayerCheckoutRadius)

This would avoid a sqrt call and instead use a less expensive sqr. 

Possibly could cache: PlayerCheckoutRadius*PlayerCheckoutRadius if it doesn't change and then you only have to do the op once.